BUG: Index.get_indexer_non_unique misbehaves with multiple nan

alexhlim opened the pull request:
- [x] closes #35392
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Looking at the implementation in index.pyx, I notice that when `[np.nan]` is passed to `get_indexer_non_unique`, the code was not able to get passed the `__contains__` method for stargets (line 314).

Further testing (python=3.7.7, numpy=1.18.5):

```
import numpy as np

# Case 1: Does not work -> prints nothing
# nan dtype: np.float64, ndarray dtype: np.float64
targets = np.array([np.nan])

# Case 2: Works -> prints 0, 1, 2
# nan dtype: U3, ndarray dtype: <U32
targets = np.array([np.nan, 'var1'])

values = np.array([np.nan, 'var1', np.nan])

stargets = set(targets)

for i, v in enumerate(values):
    if v in stargets:
        print(i)
```

Case 1 and 2 results differ because of the dtype of nan (U3 vs float64).

Upon further research, I figured out that np.nan != np.nan as per IEEE (when it is a float) and creating a set from a np.array could lead to some bizarre results (https://github.com/numpy/numpy/issues/9358). Also, since a dictionary is the main data structure in this method to keep track of the targets indices, I don't think it is ideal to use nans as keys (https://stackoverflow.com/questions/6441857/nans-as-key-in-dictionaries).

I thought it would be appropriate to replace nans (with 0s) in the targets and values arrays in order to avoid the problems stated above. When considering where to replace the nans, I thought of two places where it could potentially happen:

1. In `get_indexer_non_unique` (/pandas/core/indexes/base.py)
2. In `get_indexer_non_unique` (/pandas/_libs/index.pyx)

Including the changes in 1. would mean overwriting the the Index object's properties, so I decided to include the changes in 2.

FYI -- I wasn't sure if the test I included was in the correct file. Please let me know if you would like this test to be in another file.

arw2019 commented on pandas/_libs/index.pyx:
```diff
@@ -270,7 +270,15 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
-        values = np.array(self._get_index_values(), copy=False)
+        #  GH #35392  Index.get_indexer_non_unique misbehaves when index contains multiple nan
```
nit-pick: people usually leave comments for TO DO's. git/blame keeps track of code changes

alexhlim replied:
gotcha, removed the comment

arw2019 commented on pandas/tests/base/test_misc.py:
```diff
@@ -201,3 +201,22 @@ def test_get_indexer_non_unique_dtype_mismatch():
     tm.assert_numpy_array_equal(np.array([-1], dtype=np.intp), indexes)
     tm.assert_numpy_array_equal(np.array([0], dtype=np.int64), missing)
+
+
```
I might put it in `pandas/tests/indexes/test_base.py`, next to
 https://github.com/pandas-dev/pandas/blob/bdcc5bffaadb7488474b65554c4e8e96a00aa4af/pandas/tests/indexes/test_base.py#L2590-L2616

arw2019 reviewed (changes requested):
looks good, some comments

alexhlim commented:
Thanks for the comments, @arw2019! I just committed the changes.

arw2019 reviewed (approved):
looks good

Might want to move the release note to 1.2 - I think 1.1.1 is intended for regressions from 1.0.5

WillAyd commented on pandas/_libs/index.pyx:
```diff
@@ -270,7 +270,14 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
-        values = np.array(self._get_index_values(), copy=False)
+        if any([checknull(t) for t in targets]):
```
This already exists as a function called `isnaobj` in missing.pyx - can you use that instead?

alexhlim replied:
I changed 273 to `if isnaobj(targets).any():` and tested against the `index_cached_properties`, `index_object`, `indexing`, and `indexing_engines` benchmarks and got the following results:

```
     <master>         <nui-regression>
+         172±1μs          247±3μs     1.44  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int16Engine'>, <class 'numpy.int16'>), 'monotonic_incr')
+         188±2μs         245±30μs     1.30  index_cached_properties.IndexCache.time_is_monotonic('PeriodIndex')
+         210±2μs          237±2μs     1.13  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int8Engine'>, <class 'numpy.int8'>), 'monotonic_incr')
+      13.3±0.3μs       14.7±0.6μs     1.11  index_cached_properties.IndexCache.time_shape('CategoricalIndex')
-        11.3±2μs       10.0±0.3μs     0.89  index_cached_properties.IndexCache.time_is_all_dates('Float64Index')
-        249±30μs         215±10μs     0.86  index_cached_properties.IndexCache.time_is_monotonic('TimedeltaIndex')
-        18.8±2μs       15.9±0.9μs     0.85  index_cached_properties.IndexCache.time_engine('UInt64Index')
-      1.10±0.2ms        894±100μs     0.81  index_cached_properties.IndexCache.time_engine('MultiIndex')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

Any suggestions to improve these?

jreback replied:
you need to create a mask then you don't actually check things multiple times.

why are you creating targets/values as object arrays?

alexhlim replied:
Makes sense. `targets[isnaobj(targets)].size > 0` would work, correct? Because now instead of iterating through the array with `any()`, I'm just filtering the given array and checking its size.

I created the targets/values arrays as object arrays because I thought that it was possible to have multiple dtypes in the given arrays. Thinking about it now, I figure it makes more sense to make them as their previous array dtypes:

```
targets = np.array(new_targets, dtype=targets.dtype)
values = np.array(new_values, dtype=self._get_index_values().dtype)
```

I implemented this and the tests are passing. Does this sound good? Any suggestion to improve?

alexhlim replied:
@WillAyd I ran tests against indexing and indexing_engines benchmarks. No performance regressions:

```
asv continuous -f 1.1 upstream/master HEAD -b ^indexing_engines
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

asv continuous -f 1.1 upstream/master HEAD -b indexing
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

WillAyd reviewed (changes requested):
Can you run the indexing benchmarks to check for performance regressions?

jreback commented on pandas/tests/indexes/test_base.py:
```diff
@@ -2614,3 +2614,22 @@ def construct(dtype):
         tm.assert_numpy_array_equal(result[1], no_matches)
+
+
+@pytest.mark.parametrize(
```
this should go with other indexing tests e.g. pandas/tests/indexes/test_index.py.

can you add an object array that contains nans
can you add a DTI that has multiple NaTs

alexhlim replied:
I don't see a file called pandas/tests/indexes/test_index.py. Did you mean pandas/tests/indexes/test_indexing.py? In the file docstring, it mentions it contains tests for `get_indexer`.

alexhlim replied:
Moved and added tests

jreback commented on pandas/tests/indexes/test_base.py:
```diff
@@ -2614,3 +2614,22 @@ def construct(dtype):
+@pytest.mark.parametrize(
+    "idx, target, expected",
+    [
+        ([np.nan, "var1", np.nan], [np.nan], np.array([0, 2], dtype=np.int64)),
```
indexing with multiple nans is not very well supported at a high level, can you add a user facing example (is there on e in the OP)?

alexhlim replied:
What does OP stand for?

alexhlim replied:
Added the example from the OP in the `get_indexer_for` docs

WillAyd commented on pandas/_libs/index.pyx:
```diff
@@ -270,7 +270,14 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
-        values = np.array(self._get_index_values(), copy=False)
+        if targets[isnaobj(targets)].size > 0:
```
```suggestion
        if isnaobj(targets)].any():
```

I think the extra indexing step here is unnecessary, no?

alexhlim replied:
Agreed, I will update.

WillAyd commented on pandas/_libs/index.pyx:
```diff
@@ -270,7 +270,14 @@ cdef class IndexEngine:
         self._ensure_mapping_populated()
-        values = np.array(self._get_index_values(), copy=False)
+        if targets[isnaobj(targets)].size > 0:
+            new_targets = [0 if checknull(t) else t for t in targets]
```
Can you not just reuse the mask used for branching? So before the branch something like `mask = isnaojb(targets)` and then here do `targets[mask] = 0` ?

WillAyd replied:
Ideally want to avoid the list comprehension

alexhlim replied:
Sounds good. I will also use a mask to avoid the list comprehension in filling the values array with 0s.

WillAyd commented on pandas/tests/indexes/test_indexing.py:
```diff
@@ -96,3 +99,31 @@ def test_getitem_deprecated_float(idx):
+@pytest.mark.parametrize(
+    "idx, target, expected",
+    [
+        ([np.nan, "var1", np.nan], [np.nan], np.array([0, 2], dtype=np.int64)),
```
Can you add a test case that includes both np.nan and 0 to make sure this can disambiguate?

alexhlim replied:
Good point, I don't think this will differentiate between the two because the conversion from np.nan to 0 happens before the check. I have a few of ideas:
1) Should we choose another replace value for np.nan with (ie. 'NaN')?
2) Check to see if targets/values have 0s in it and replace it with something else?
Let me know what direction you think I should go? Thanks!

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -275,6 +275,12 @@ cdef class IndexEngine:
+        targets_mask = isnaobj(targets)
+        values_mask = isnaobj(values)
+        if targets_mask.any():
+            targets[targets_mask] = 0
```
we are mutating the inputs, *might* be ok, but can copy before you do this is better

alexhlim replied:
gotcha, I made a copy of values as well (since it is not a copy and we are modifying it)

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -275,6 +275,12 @@ cdef class IndexEngine:
+        values_mask = isnaobj(values)
+        if targets_mask.any():
+            targets[targets_mask] = 0
+            values[values_mask] = 0
```
why are you setting to 0? these are indexers so shouldn't these be -1 ?

alexhlim replied:
agreed, changed to -1

jreback commented on pandas/tests/indexes/test_indexing.py:
```diff
@@ -96,3 +99,46 @@ def test_getitem_deprecated_float(idx):
+@pytest.mark.parametrize(
+    "idx, target, expected",
+    [
+        ([np.nan, "var1", np.nan], [np.nan], np.array([0, 2], dtype=np.int64)),
```
~all of the expected should be np.intp~ wow we actually return np.int64, this is totally wrong, these *should* all be `np.intp` (but dont' change here, can you raise an issue for this).

alexhlim replied:
raised new issue below

jreback reviewed (changes requested):
can you merge master and a few comments.

jbrockmendel commented on pandas/tests/indexes/test_indexing.py:
```diff
@@ -16,7 +17,9 @@
 import pytest
 
-from pandas import Float64Index, Index, Int64Index, UInt64Index
+from pandas._libs.tslibs import NaT
```
this can just go in the `from pandas import` line

jbrockmendel commented on pandas/tests/indexes/test_indexing.py:
```diff
@@ -96,3 +99,46 @@ def test_getitem_deprecated_float(idx):
+@pytest.mark.parametrize(
+    "idx, target, expected",
+    [
+        ([np.nan, "var1", np.nan], [np.nan], np.array([0, 2], dtype=np.intp)),
```
does np.nan vs `float("NaN")` make a difference?

alexhlim replied:
I don't believe so. I changed all the np.nan to float("NaN") and tests are all passing. Would you like me to include a separate test case for float("NaN")?

arw2019 replied:
please add that testcase

arw2019 reviewed (commented):
@alexhlim is this still active? If yes can you merge master and address review comments?

jbrockmendel commented:
Is it only/mainly object-dtype Index that is getting this wrong?

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -275,6 +275,14 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
         values = np.array(self._get_index_values(), copy=False)
+        targets_mask = isnaobj(targets)
```
any particular reason you're doing this up-front instead of in the loop below?

alexhlim replied:
I did this up-front because when you store NaN in a stargets (line 286), the `val in stargets` (line 318) statement does not work

alexhlim commented:
@jbrockmendel tested this on master and it looks like even when using Float64Index, it is incorrect:
```
>>> import pandas as pd
>>> import numpy as np
>>> axis = pd.Index([np.nan, 1.0, np.nan], dtype=np.float64)
>>> axis.get_indexer_for([np.nan])
array([-1])
>>> axis
Float64Index([nan, 1.0, nan], dtype='float64')
```

alexhlim commented:
@jbrockmendel - my PR caused your test to fail:

In pandas/tests/indexing/test_loc.py (test_loc_getitem_listlike_unused_category_raises_keyerro) are raising the following errors:
```
KeyError: "Passing list-likes to .loc or [] with any missing labels is no longer supported. The following labels were missing: Index([-1], dtype='object'). See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"
```

```
AssertionError: Regex pattern 'Passing list-likes to .loc or \\[\\] with any missing labels is no longer supported. The following labels were missing: (Categorical)?Index\\(\\[nan\\], .*\\). See https' does not match '"Passing list-likes to .loc or [] with any missing labels is no longer supported. The following labels were missing: Index([-1], dtype=\'object\'). See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"'.
```

Since I am setting the NaN values in the Index to -1, it is causing `_validate_read_indexer` to trigger its assertion statement. I'm unsure how to resolve this. Any ideas?

arw2019 commented on pandas/_libs/index.pyx:
```diff
@@ -275,7 +275,14 @@ cdef class IndexEngine:
+        values_mask = isnaobj(values)
+        targets_mask = isnaobj(targets)
+        if targets_mask.any():
+            targets_no_nan = targets.copy()
```
is it necessary to copy? could you do `stargets=set(targets)` and `stargets.add(-1)`

alexhlim replied:
good call. I can add to set instead

arw2019 commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -697,8 +697,13 @@ Indexing
 - Bug in :meth:`DataFrame.loc` and :meth:`DataFrame.__getitem__`  raising ``KeyError`` when columns were :class:`MultiIndex` with only one level (:issue:`29749`)
 - Bug in :meth:`Series.__getitem__` and :meth:`DataFrame.__getitem__` raising blank ``KeyError`` without missing keys for :class:`IntervalIndex` (:issue:`27365`)
 - Bug in setting a new label on a :class:`DataFrame` or :class:`Series` with a :class:`CategoricalIndex` incorrectly raising ``TypeError`` when the new label is not among the index's categories (:issue:`38098`)
+<<<<<<< HEAD
```
merge error here

jbrockmendel commented:
> Since I am setting the NaN values in the Index to -1, it is causing _validate_read_indexer to trigger its assertion statement. I'm unsure how to resolve this. Is my implementation not supportable due to this error?

The user-facing exception message should reflect the key the user actually passed.  So IIUC you need to sort of undo the np.nan->0 replacement

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -324,8 +336,19 @@ cdef class IndexEngine:
 
-                    result[count] = j
-                    count += 1
+                    if val == -1:
```
can you add comments explaining whats going on here, or at least pointing back to this PR/issue

alexhlim replied:
will do

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -275,7 +275,11 @@ cdef class IndexEngine:
+        targets_mask = isnaobj(targets)
         stargets = set(targets)
+        if targets_mask.any():
+            stargets.add(-1)
```
do we know that -1 isnt already in stargets?  or does it not matter?

alexhlim replied:
I can revise to `targets_mask.any() and -1 not in stargets`. however, I don't think it matters because the end result will be the same

arw2019 replied:
Well if we know then we can consolidate this whole if else no?

alexhlim replied:
@arw2019 by consolidate, do you mean checking for -1 in the set before checking for nans? like what I wrote above?

arw2019 replied:
As long as -1 doesn't need extra handling we can just delete this if statement 

I see you already consolidated

alexhlim replied:
I think we'll still need this if statement. consider the following scenario:

```
>>> import numpy as np
>>> targets = [np.nan]
>>> stargets = set(targets)
>>> stargets
{nan}
```

we would need to add -1 in stargets to ensure this line works:

```
>>> val = -1
>>> if val in stargets:
...     print("in stargets")
... 
```

alexhlim replied:
I kept the if statement and added a check for -1 in my most recent commit - does this work?

arw2019 replied:
yeah i think so

alexhlim commented:
> > Since I am setting the NaN values in the Index to -1, it is causing _validate_read_indexer to trigger its assertion statement. I'm unsure how to resolve this. Is my implementation not supportable due to this error?
> 
> The user-facing exception message should reflect the key the user actually passed. So IIUC you need to sort of undo the np.nan->0 replacement

agreed. my new implementation does not mutate the original target/values, and instead handles within the for loop.

arw2019 commented on pandas/tests/indexes/test_indexing.py:
```diff
@@ -148,3 +150,51 @@ def test_getitem_deprecated_float(idx):
+)
+def test_get_indexer_non_unique_multiple_nans(idx, target, expected):
+    # GH 35392
+    axis = Index(idx)
```
total nit but i'd combine these two lines into 
``` python
result = Index(axis).get_indexer_for(target)
```

alexhlim replied:
agreed we can combine - I'll change

jreback commented:
lgtm. @WillAyd or @jbrockmendel if any other comments.

simonjayhawkins commented:
@alexhlim can you merge master to resolve conflicts.

alexhlim commented:
@simonjayhawkins just merged with master!

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -278,7 +281,11 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
         values = np.array(self._get_index_values(), copy=False)
+        values_mask = isnaobj(values)
```
i dont think values is necessary object dtype is it?  (if it is, should make declaration more specific)

alexhlim replied:
@jbrockmendel - you're correct, values is not necessarily an object dtype. in my tests, the following are not object dtype:

1)`DatetimeIndex(['2020-08-05', 'NaT', 'NaT'], dtype='datetime64[ns]', freq=None)` -> int64
2) `array([nan, -1, nan], dtype=object)` -> float64

Knowing this, should anything be changed from your perspective?

jbrockmendel replied:
we'd want to use a dtype-specific `isna` function.  usually that would be pd.isna, but in this file we try pretty hard to avoid importing from outside of our cython code.  Before I can make a suggestion on dealing with that, I need to ask about the dt64 case.

> `DatetimeIndex(['2020-08-05', 'NaT', 'NaT'], dtype='datetime64[ns]', freq=None) -> int64`

IIUC you're `values` here would be `np.array([ 1596585600000000000, -9223372036854775808, -9223372036854775808], dtype=np.int64)`.  Then values_mask would be all-False, despite the fact that the index contains NaTs.  Is that intentional?

If so, then wouldnt it make sense, to override (part of) this method for IndexEngine subclasses where values_mask will always be all-False?

alexhlim replied:
That was not intentional. I was under the assumption that `isnaobj` would capture the `NaT` in values_mask as True. Now, from what I understand, `isnaobj` only detects NA-like values when the array dtype=object. Is this correct?

 Interestingly, the dt64 case still passed because both values of NaT were the same integer, so keys were properly updated in `d` (dict): `{-9223372036854775808: [1, 2]}`

I probably shouldn't be relying on the above case and instead use a dt64 na checker.

Could you help point out to me which functions I should be using to check NA for dt64 and float64? Are there other dtypes I should check?

jbrockmendel replied:
> Now, from what I understand, isnaobj only detects NA-like values when the array dtype=object. Is this correct?

Not quite.  If you pass a ndarray[float64] it will correctly identify NaNs, same with ndarray[datetime64ns] and NaTs.  In this case you're just passing int64, which isnaobj will always come give all-False.

> Could you help point out to me which functions I should be using to check NA for dt64 and float64? Are there other dtypes I should check?

IIUC then the check you are implementing is only relevant if `values` can contain NaN, i.e. only floating or complex dtypes.  So I'd suggest 1) implementing an override that skips the new checks for other dtypes and 2) just using `np.isnan(values)`

jreback commented:
@alexhlim ok i think @jbrockmendel has a question. Are all the cases in the OP covered by tests? (meaning the user facing cases). can you merge master, ping when ready.

alexhlim commented:
@jreback all cases in OP are in included in the tests. once I resolve issue above, I will merge master and ping.

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -389,6 +414,13 @@ cdef class ObjectEngine(IndexEngine):
         return _hash.PyObjectHashTable(n)
 
+    def get_stargets(self, ndarray targets) -> set:
+        stargets = set(targets)
```
can you add a doc-string here (and above)

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -267,6 +271,25 @@ cdef class IndexEngine:
+            return val
+
+    def should_update_d(self, object target, object val) -> bool:
+        # -1 in targets could be either -1 or nan
```
can you add a full doc-string (e.g. Parameters / Returns), came for convert_val_if_nan

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -267,6 +271,25 @@ cdef class IndexEngine:
+        return set(targets)
+
+    def convert_val_if_nan(self, object val) -> object:
+        # unable to utilize val if nan when updating
```
i would just inline this where you need ti

jreback reviewed (changes requested):
pls rebase as well

jreback commented:
@alexhlim can you also run the indexing perf asv's for this.

jbrockmendel commented:
Is the issue multiple NaNs, or would this also fix `pd.Index(["a", "b", "a", np.nan]).get_indexer_non_unique([np.nan])`?

jbrockmendel commented on pandas/_libs/index_class_helper.pxi.in:
```diff
@@ -77,4 +77,14 @@ cdef class {{name}}Engine(IndexEngine):
+        stargets = set(targets)
+        # account for nans
+        if -1 not in stargets and np.isnan(targets).any():
+            stargets.add(-1)
```
why are you using -1 instead of just a canonical np.nan object?

realead replied:
I like the idea of using a canonical nan-object more than, replacing a nan by -1. Because, what happens if we have nans and -1s? This uses the fact, that python first compares the identity of objects and only then the objects, so `np.nan == np.nan` because it is the same object.

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -267,6 +271,25 @@ cdef class IndexEngine:
+        else:
+            return val
+
+    def should_update_d(self, object target, object val) -> bool:
```
i think this can be subclass-specific too, only ObjectEngine and Float(64|32)Engine need to check for NAs

jbrockmendel commented on pandas/_libs/index_class_helper.pxi.in:
```diff
@@ -77,4 +77,14 @@ cdef class {{name}}Engine(IndexEngine):
+    def get_stargets(self, ndarray targets):
+        stargets = set(targets)
+        # account for nans
+        if -1 not in stargets and np.isnan(targets).any():
```
we can improve on this slightly by first doing a (cached) check for if any of the engine's values are nan.  if not, then we dont need to adjust stargets

alexhlim commented:
@jbrockmendel I noticed you submitted a PR that will close the original issue. if that's the case, would you like me to close this PR? or are there some changes that I have incorporated in this PR that you would like merged?

jbrockmendel commented:
>  I noticed you submitted a PR that will close the original issue. if that's the case, would you like me to close this PR? or are there some changes that I have incorporated in this PR that you would like merged?

I was hoping you'd port whatever parts of my PR are useful and then I'll close mine.  You did most of the work here.

alexhlim commented:
> > I noticed you submitted a PR that will close the original issue. if that's the case, would you like me to close this PR? or are there some changes that I have incorporated in this PR that you would like merged?
> 
> I was hoping you'd port whatever parts of my PR are useful and then I'll close mine. You did most of the work here.

Thanks for the guidance, I will definitely port your work into this PR!

alexhlim committed:
BUG: Index.get_indexer_non_unique misbehaves when index contains multiple nan (#35392)

alexhlim commented:
@jbrockmendel I liked the simplicity of your implementation better, so I decided to use it instead of what I currently have. I added a test case for the matching-but-not-identical nans. also, I rebased this PR so there's a cleaner commit history.

alexhlim commented:
@jreback ran the index asv: some performance decreased:

```
asv continuous -f 1.1 upstream/master HEAD -b ^indexing

       before           after         ratio
     [5675cd8a]       [85dd25ce]
     <master>         <nui-regression>
+        81.5±2ms          107±2ms     1.31  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'nonunique_monotonic_inc')
+        80.4±1ms          105±3ms     1.31  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
+        81.8±2ms          106±1ms     1.29  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
+      83.3±0.9ms        105±0.9ms     1.26  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'nonunique_monotonic_inc')
+        81.9±2ms          101±2ms     1.24  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
+      8.14±0.3ms       10.0±0.2ms     1.23  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'nonunique_monotonic_inc')
+        88.4±2ms          106±2ms     1.19  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+        88.5±1ms          106±1ms     1.19  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+         128±2μs          143±3μs     1.12  indexing.NumericSeriesIndexing.time_getitem_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.

```

jbrockmendel commented:
LGTM @alexhlim can you merge master

alexhlim committed:
merging master

jbrockmendel commented:
LGTM pending green cc @jreback

alexhlim commented:
@jbrockmendel noticed that CI / Checks is failing:

1. [doctest] pandas.core.generic.NDFrame.to_xarray
2. Typing validation

Are these errors something that my PR caused?  If not, should I wait until these errors are resolved on master and re-merge?

jbrockmendel commented:
Looks like those are unrelated, #42716

alexhlim committed:
merging master

alexhlim commented:
@jbrockmendel @jreback pinging b/c green

jbrockmendel reviewed (approved):
LGTM cc @jreback

jreback committed:
Merge branch 'master' into nui-regression

jreback commented:
thanks @alexhlim very nice, will merge on green.

jreback commented:
thanks @alexhlim

alexhlim commented:
Thank you everyone, especially @jbrockmendel and @jreback for guiding me through this PR. I definitely learned a lot!
